storage: clean up metrics

tamird opened the pull request:
I'll open a master -> develop after this guy since there'll be conflict.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/8861)

<!-- Reviewable:end -->

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 3 of 3 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[storage/replica.go, line 1766 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8861#-KQ96H1AOvesLZJc179a:-KQ96H1AOvesLZJc179b:bo1n41s) ([raw file](https://github.com/cockroachdb/cockroach/blob/421c52e69e8e028739576fac23d76ed892ff43a1/storage/replica.go#L1766)):*
> ```Go
> // tick the Raft group, returning any error and true if the raft group exists
> // and false otherwise. Updates acc based on the replica's state.
> func (r *Replica) tick(acc *tickAccumulator) (bool, error) {
> ```

The `bool` is

---

*[storage/store.go, line 2576 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8861#-KQ96VITXhLEniOGus-3:-KQ96VITXhLEniOGus-4:byz1j1h) ([raw file](https://github.com/cockroachdb/cockroach/blob/421c52e69e8e028739576fac23d76ed892ff43a1/storage/store.go#L2576)):*
> ```Go
> 				pendingReplicas = pendingReplicas[:0]
> 
> 				var acc tickAccumulator
> ```

you could add `pendingReplicas` into tickInfo and avoid the somewhat awkward `bool` return. Your call.

Also, `len(pendingReplicas)` is the number of non-quiesced Raft groups. I don't know if we capture that elsewhere, but if we don't we should (though it isn't as pressing yet since raft groups are only rarely destroyed once they exist).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8861#-:-KQ96slGbSi-ePAYnaKs:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[storage/replica.go, line 1766 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8861#-KQ96H1AOvesLZJc179a:-KQ976i8pq_vEsrsy9KH:b5zxzs3) ([raw file](https://github.com/cockroachdb/cockroach/blob/421c52e69e8e028739576fac23d76ed892ff43a1/storage/replica.go#L1766)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> The `bool` is
</details>
didn't finish the thought?

---

*[storage/store.go, line 2576 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8861#-KQ96VITXhLEniOGus-3:-KQ99ULfkY0mTBsMrbWE:bv3qk9w) ([raw file](https://github.com/cockroachdb/cockroach/blob/421c52e69e8e028739576fac23d76ed892ff43a1/storage/store.go#L2576)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> you could add `pendingReplicas` into tickInfo and avoid the somewhat awkward `bool` return. Your call.
> 
> Also, `len(pendingReplicas)` is the number of non-quiesced Raft groups. I don't know if we capture that elsewhere, but if we don't we should (though it isn't as pressing yet since raft groups are only rarely destroyed once they exist).
</details>
re `pendingReplicas`: done. I'll add len(pendingReplicas) and the transport stats tomorrow.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8861)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 3 of 3 files at r3, 3 of 3 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[storage/replica.go, line 1766 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8861#-KQ96H1AOvesLZJc179a:-KQ9NzF6gL22sY5Dq1UV:b-qimc9j) ([raw file](https://github.com/cockroachdb/cockroach/blob/421c52e69e8e028739576fac23d76ed892ff43a1/storage/replica.go#L1766)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> didn't finish the thought?
</details>
Sorta. It's finished it in the other comment.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8861)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Mind adding

```go
diff --git a/ui/app/containers/nodeGraphs.tsx b/ui/app/containers/nodeGraphs.tsx
index 71d382d..084bfa0 100644
--- a/ui/app/containers/nodeGraphs.tsx
+++ b/ui/app/containers/nodeGraphs.tsx
@@ -242,9 +242,9 @@ export default class extends React.Component<IInjectedProps, {}> {

             <LineGraph title="Raft Leaders and LeaseHolders" sources={sources}>
               <Axis format={ d3.format(".1f") }>
-                <Metric name="cr.store.raft.leaders" title="Ranges on this Store that are Raft Leaders" />
-                <Metric name="cr.store.range.leaseholders" title="Ranges on this Store that are Raft LeaseHolders" />
-                <Metric name="cr.store.range.leaseholders.without.leadership" title="Ranges on this Store that are Raft LeaseHolders but aren't Raft Leaders" />
+                <Metric name="cr.store.raft.leaders" title="Raft leaders Replicas" />
+                <Metric name="cr.store.range.leaseholders" title="Leaseholder Replicas" />
+                <Metric name="cr.store.range.leaseholders.without.leadership" title="Not Leader, but LeaseHolder" />
               </Axis>
             </LineGraph>

```

in (might not apply cleanly with your changes). The original text gets truncated and only hovering for a few seconds returns the true meaning - no bueno.

tamird committed:
storage: use anonymous struct to clarify locking

tamird committed:
storage: simplify metrics updates

Removed methods on storeMetrics that do nothing but indirect.

tamird committed:
storage: clean up metrics

Move the work from (*Replica).tick() to (*Store).ComputeMetrics().

Also replace "fully replicated" and "replication pending" into more
specific metrics.

tamird commented:
big overhaul here. PTAL

tschottdorf commented:
Reviewed 6 of 6 files at r5, 2 of 2 files at r6, 12 of 12 files at r7.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[storage/store.go, line 2809 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8861#-KQbMl3VKvRFWDKdlioB:-KQbMl3VKvRFWDKdlioC:bfwp2ng) ([raw file](https://github.com/cockroachdb/cockroach/blob/ceb5364c15eda9547b87f8e54e80d3b436f43c6d/storage/store.go#L2809)):*
> ```Go
> 		if raftStatus != nil && raftStatus.SoftState.RaftState == raft.StateLeader {
> 			lease := rng.mu.state.Lease
> 			rng.mu.Unlock()
> ```

What's with this odd locking?

---

*[storage/store_metrics.go, line 157 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8861#-KQbN5PNoZyN8uKkA7cM:-KQbN5PNoZyN8uKkA7cN:b-r52v0l) ([raw file](https://github.com/cockroachdb/cockroach/blob/506e48097be8c5ef1bc6b21b941d67192650024d/storage/store_metrics.go#L157)):*
> ```Go
> 	ReservedReplicaCount         *metric.Counter
> 	LeaderRangeCount             *metric.Gauge
> 	ReplicatedRangeCount         *metric.Gauge
> ```

What was ReplicatedRangeCount counting exactly?

---

*[storage/store_metrics.go, line 32 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8861#-KQbMtAu9DEmwFBVSCwe:-KQbMtAu9DEmwFBVSCwf:bic2g3j) ([raw file](https://github.com/cockroachdb/cockroach/blob/ceb5364c15eda9547b87f8e54e80d3b436f43c6d/storage/store_metrics.go#L32)):*
> ```Go
> 	metaReservedReplicaCount          = metric.Metadata{Name: "replicas.reserved"}
> 	metaRaftLeaderCount               = metric.Metadata{Name: "replicas.leaders"}
> 	metaRaftLeaderNotLeaseHolderCount = metric.Metadata{Name: "replicas.leaders.not.leaseholders"}
> ```

Is `leaders.not.leaseholders` a good idea?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8861)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[storage/store.go, line 2809 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8861#-KQbMl3VKvRFWDKdlioB:-KQbO1L0ixFpVcmmiXcw:bj7eqmo) ([raw file](https://github.com/cockroachdb/cockroach/blob/ceb5364c15eda9547b87f8e54e80d3b436f43c6d/storage/store.go#L2809)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> What's with this odd locking?
</details>
grabbing the lease only if needed. I can simplify if you prefer.

---

*[storage/store_metrics.go, line 157 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8861#-KQbN5PNoZyN8uKkA7cM:-KQbNik_IOVXFYCNCvBo:b-dzwgdo) ([raw file](https://github.com/cockroachdb/cockroach/blob/506e48097be8c5ef1bc6b21b941d67192650024d/storage/store_metrics.go#L157)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> What was ReplicatedRangeCount counting exactly?
</details>
It was counting ranges which had the ZoneConfig-specified number of replicas. I didn't think that stat was useful, which is why I replaced it with the replication metrics below. I could be convinced otherwise.

---

*[storage/store_metrics.go, line 32 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8861#-KQbMtAu9DEmwFBVSCwe:-KQbNZaO0-I428kdo3V7:bx4z241) ([raw file](https://github.com/cockroachdb/cockroach/blob/ceb5364c15eda9547b87f8e54e80d3b436f43c6d/storage/store_metrics.go#L32)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Is `leaders.not.leaseholders` a good idea?
</details>
Can you say more words? What would you prefer?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8861)*
<!-- Sent from Reviewable.io -->

mberhault commented on storage/store_metrics.go:
```diff
@@ -26,30 +25,43 @@ import (
-	metaReserved                     = metric.Metadata{Name: "capacity.reserved"}
-	metaSysBytes                     = metric.Metadata{Name: "sysbytes"}
-	metaSysCount                     = metric.Metadata{Name: "syscount"}
+	// Replica metrics.
```
since you're renaming a bunch of metrics, you should go ahead and fill in the help strings

tschottdorf commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[storage/store.go, line 2809 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8861#-KQbMl3VKvRFWDKdlioB:-KQbRF9mSJ81XZ9uyrer:b-gt5s8k) ([raw file](https://github.com/cockroachdb/cockroach/blob/ceb5364c15eda9547b87f8e54e80d3b436f43c6d/storage/store.go#L2809)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> grabbing the lease only if needed. I can simplify if you prefer.
</details>
You could make all of this conditional on `lease != nil` and set the lease only if we're the leader. Perhaps that would reduce the lock to grabbing the raftStatus and then perhaps the lease, and nothing more?

---

*[storage/store_metrics.go, line 157 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8861#-KQbN5PNoZyN8uKkA7cM:-KQbQtsULMxujs_Sa-ia:b-svtjuh) ([raw file](https://github.com/cockroachdb/cockroach/blob/506e48097be8c5ef1bc6b21b941d67192650024d/storage/store_metrics.go#L157)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> It was counting ranges which had the ZoneConfig-specified number of replicas. I didn't think that stat was useful, which is why I replaced it with the replication metrics below. I could be convinced otherwise.
</details>
Seems more useful to me the way it is now.

---

*[storage/store_metrics.go, line 28 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8861#-KQbHH00krJ8RP56bnse-r7-28:-KQbRV53th7IM1TaetlK:bxd63z2) ([raw file](https://github.com/cockroachdb/cockroach/blob/ceb5364c15eda9547b87f8e54e80d3b436f43c6d/storage/store_metrics.go#L28)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary>
> since you're renaming a bunch of metrics, you should go ahead and fill in the help strings
</details>
Yeah, and compile a shortlist of changes to update the grafana stuff if needed.

---

*[storage/store_metrics.go, line 32 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8861#-KQbMtAu9DEmwFBVSCwe:-KQbR0jhd0xrwMJ7LFrT:bog32hk) ([raw file](https://github.com/cockroachdb/cockroach/blob/ceb5364c15eda9547b87f8e54e80d3b436f43c6d/storage/store_metrics.go#L32)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Can you say more words? What would you prefer?
</details>
thinking of these names as forming a tree, I'd have expected it next to `replicas.leaders` (i.e. `replicas.leaders_not_leaseholders`).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8861)*
<!-- Sent from Reviewable.io -->

mberhault commented on storage/store_metrics.go:
```diff
@@ -26,30 +25,43 @@ import (
+	metaReplicaCount                  = metric.Metadata{Name: "replicas"}
+	metaReservedReplicaCount          = metric.Metadata{Name: "replicas.reserved"}
+	metaRaftLeaderCount               = metric.Metadata{Name: "replicas.leaders"}
+	metaRaftLeaderNotLeaseHolderCount = metric.Metadata{Name: "replicas.leaders.not.leaseholders"}
```
be aware that exported metrics will have all characters other than `[_a-zA-Z0-9]` converted to underscores, making names like this one rather tedious to read

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>. I also asked Arjun to take a look.

---

Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8861#-:-KQbRa4Q5NLn2tcAz9vk:b-242d7d)*
<!-- Sent from Reviewable.io -->

arjunravinarayan commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[storage/store.go, line 2648 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8861#-KQbQrzKm8ed1A6Uwmuc:-KQbQrzKm8ed1A6Uwmud:b-bksdpb) ([raw file](https://github.com/cockroachdb/cockroach/blob/f05aa3f76d4a3fd3366643e4c66493c90a60d483/storage/store.go#L2648)):*
> ```Go
> 						pendingReplicas = append(pendingReplicas, id)
> 					}
> 					if info.IsRangeLeaseHolder && !info.IsRaftLeader {
> ```

Stub comment pointing out that the old behavior was to track *lease holders who weren't raft leaders* (see other comment for context).

---

*[storage/store.go, line 2820 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8861#-KQbR3ye5kGsvAXmQiMm:-KQbR3ye5kGsvAXmQiMn:b-xpwq6) ([raw file](https://github.com/cockroachdb/cockroach/blob/ceb5364c15eda9547b87f8e54e80d3b436f43c6d/storage/store.go#L2820)):*
> ```Go
> 					leaseHolderCount++
> 				} else {
> 					raftLeaderNotLeaseHolderCount++
> ```

So you've swapped the behavior here. Here you're counting replicas that are raft leaders but not lease holders. IDK if this is important, but I put it way it was because that's how @andreimatei requested the stats. Perhaps it's not important either way (and given that you're only tracking metrics on raft leaders I can see how this way is simpler), but just want to know what your reasoning.

---

*[storage/store_metrics.go, line 32 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8861#-KQbMtAu9DEmwFBVSCwe:-KQbPF2odWoZRIfWnyB_:b-tudgtk) ([raw file](https://github.com/cockroachdb/cockroach/blob/ceb5364c15eda9547b87f8e54e80d3b436f43c6d/storage/store_metrics.go#L32)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> thinking of these names as forming a tree, I'd have expected it next to `replicas.leaders` (i.e. `replicas.leaders_not_leaseholders`).
</details>
(Stepping in for Toby) `replica.leaders.who.arent.range.leaseholders`?  Perhaps just add a comment describing the exact behavior, this variable is too long as it is.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8861#-:-KQbNnrHD0Hs-KnVKPYi:b-jbsf8z)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[storage/store.go, line 2809 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8861#-KQbMl3VKvRFWDKdlioB:-KQbSQrxKAKor5iMY9Ad:b-2ribsh) ([raw file](https://github.com/cockroachdb/cockroach/blob/ceb5364c15eda9547b87f8e54e80d3b436f43c6d/storage/store.go#L2809)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> You could make all of this conditional on `lease != nil` and set the lease only if we're the leader. Perhaps that would reduce the lock to grabbing the raftStatus and then perhaps the lease, and nothing more?
</details>
Made this "trivial".

---

*[storage/store.go, line 2820 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8861#-KQbR3ye5kGsvAXmQiMm:-KQbSUX79UldRtJejtPa:bt4tfup) ([raw file](https://github.com/cockroachdb/cockroach/blob/ceb5364c15eda9547b87f8e54e80d3b436f43c6d/storage/store.go#L2820)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary>
> So you've swapped the behavior here. Here you're counting replicas that are raft leaders but not lease holders. IDK if this is important, but I put it way it was because that's how @andreimatei requested the stats. Perhaps it's not important either way (and given that you're only tracking metrics on raft leaders I can see how this way is simpler), but just want to know what your reasoning.
</details>
It made the code here simpler. I discussed with @tschottdorf and confirmed it was equivalent.

---

*[storage/store_metrics.go, line 157 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8861#-KQbN5PNoZyN8uKkA7cM:-KQbWoHSmGEV8xlGuvhY:b-t7umkb) ([raw file](https://github.com/cockroachdb/cockroach/blob/506e48097be8c5ef1bc6b21b941d67192650024d/storage/store_metrics.go#L157)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Seems more useful to me the way it is now.
</details>
It does? The UI now displays something more useful (to me): "how many ranges are considered by the allocator to be OK"

---

*[storage/store_metrics.go, line 28 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8861#-KQbHH00krJ8RP56bnse-r7-28:-KQbXP_qcm6H2rYBVoo4:b7wwz0g) ([raw file](https://github.com/cockroachdb/cockroach/blob/ceb5364c15eda9547b87f8e54e80d3b436f43c6d/storage/store_metrics.go#L28)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Yeah, and compile a shortlist of changes to update the grafana stuff if needed.
</details>
s/ranges.leaders/replicas.leaders/
s/ranges.replicated/ranges.allocator.noop/

ranges.replication-pending -> ranges.allocator.{remove,add,removedead} (doesn't have an exact corollary)

s/raft.leaders/replicas.leaders/
s/range.leaseholders/replicas.leaseholders/
s/range.leaseholders.without.leadership/replicas.leaders_not_leaseholders/

I don't want to fill in all the help strings just so they're filled in. Are there any names here in particular that require explanation? I've added one for `replicas.leaders_not_leaseholders`.

---

*[storage/store_metrics.go, line 32 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8861#-KQbHH00krJ8RP56bnse-r7-32:-KQbUczwRkvfNIWllVYy:bfbnma4) ([raw file](https://github.com/cockroachdb/cockroach/blob/ceb5364c15eda9547b87f8e54e80d3b436f43c6d/storage/store_metrics.go#L32)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary>
> be aware that exported metrics will have all characters other than `[_a-zA-Z0-9]` converted to underscores, making names like this one rather tedious to read
</details>
Happy to take suggestions!

@tschottdorf I was thinking this was a subset of `replicas.leaders`, but I'll take your suggestion since nothing better has emerged.

---

*[storage/store_metrics.go, line 32 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8861#-KQbMtAu9DEmwFBVSCwe:-KQbUnjGrfycKortCp6v:bafjas2) ([raw file](https://github.com/cockroachdb/cockroach/blob/ceb5364c15eda9547b87f8e54e80d3b436f43c6d/storage/store_metrics.go#L32)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary>
> (Stepping in for Toby) `replica.leaders.who.arent.range.leaseholders`?  Perhaps just add a comment describing the exact behavior, this variable is too long as it is.
</details>
Let's move this discussion under @mberhault's comment.

---

*[ui/app/util/proto.ts, line 44 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8861#-KQbX3kJ4G8pOWpIWb1L:-KQbX3kJ4G8pOWpIWb1M:bj7swrr) ([raw file](https://github.com/cockroachdb/cockroach/blob/ceb5364c15eda9547b87f8e54e80d3b436f43c6d/ui/app/util/proto.ts#L44)):*
> ```XML
>   export const leaseHolders: string = "replicas.leaseholders";
>   export const availableRanges: string = "ranges.available";
>   export const replicatedRanges: string  = "ranges.allocator.noop";
> ```

@tschottdorf stub comment showing the new meaning of "replicatedRanges" in the UI

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8861)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 12 of 12 files at r8, 2 of 2 files at r9, 12 of 12 files at r10.
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[storage/store_metrics.go, line 157 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8861#-KQbN5PNoZyN8uKkA7cM:-KQbihQTsy4PoVT7SvdR:bg938mp) ([raw file](https://github.com/cockroachdb/cockroach/blob/506e48097be8c5ef1bc6b21b941d67192650024d/storage/store_metrics.go#L157)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> It does? The UI now displays something more useful (to me): "how many ranges are considered by the allocator to be OK"
</details>
Sorry, that came out wrong. "Now" is "in this change".

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8861)*
<!-- Sent from Reviewable.io -->

petermattis commented:
This is exactly the sort of change we shouldn't have rushed into master given how it has produced irritating conflicts with #8941. I'm not commenting on the merits of the change, just noting that it is not improving stability or addressing a major stability problem. I'm tempted to revert it.

tamird commented:
This change was made as a follow up to #8803, and does expose better
replication metrics (which may be useful for stability work given #9031).

Regarding the irritating conflicts: that's not really something that can
easily be avoided. For instance, #8956 has produced quite irritating
conflicts with #8726 and #8958, but what can we do? Someone's going to have
to resolve the conflicts somewhere.

On Thu, Sep 1, 2016 at 8:48 PM, Peter Mattis <notifications@github.com>
wrote:

> This is exactly the sort of change we shouldn't have rushed into master
> given how it has produced irritating conflicts with #8941
> <https://github.com/cockroachdb/cockroach/pull/8941>. I'm not commenting
> on the merits of the change, just noting that it is not improving stability
> or addressing a major stability problem. I'm tempted to revert it.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/8861#issuecomment-244254585>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABdsPIl8x8lo_gsibdsJjQbqMrAYpYY1ks5ql3J2gaJpZM4JumSU>
> .
>

petermattis commented:
Selfishly, I think the responsibility of handling the conflicts should fall to the author of the less risky/intricate change. Note that #8956 was a prerequisite for #8941 and, in my estimation, the snapshot death spiral that #8941 is working towards is the most pressing stability issue. 

That said, I've got the conflicts resolved. Consider the above venting, but let's be a little slower to jam things into master in the future. I was so heads down on #8941 today that I didn't even have a chance to look at this change before it merged.

bdarnell commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[storage/store.go, line 2820 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8861#-KQbR3ye5kGsvAXmQiMm:-KQeCGklYFUPFUfCkX0y:bbmirxr) ([raw file](https://github.com/cockroachdb/cockroach/blob/ceb5364c15eda9547b87f8e54e80d3b436f43c6d/storage/store.go#L2820)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> It made the code here simpler. I discussed with @tschottdorf and confirmed it was equivalent.
</details>
I don't think it's equivalent. Currently, the lease can become quiescent but raft leadership cannot, so it's normal for there to be raft leaders with no corresponding lease holder, but the reverse is not true and that's what we're trying to monitor. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8861)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[cli/node.go, line 92 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8861#-KQeD2E7jIXJaVTDHUmC:-KQeD2E8jvC-nj1EjzEE:b-qghx3n) ([raw file](https://github.com/cockroachdb/cockroach/blob/94a8c6a8797168c0500ae65014127da0799a45ee/cli/node.go#L92)):*
> ```Go
> 	"replicas_leaders",
> 	"replicas_leaseholders",
> 	"ranges_availalbe",
> ```

s/lb/bl/

---

*[storage/store.go, line 2820 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8861#-KQbR3ye5kGsvAXmQiMm:-KQeDwv0Qrl52msVNxXr:b-7h2chx) ([raw file](https://github.com/cockroachdb/cockroach/blob/ceb5364c15eda9547b87f8e54e80d3b436f43c6d/storage/store.go#L2820)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> I don't think it's equivalent. Currently, the lease can become quiescent but raft leadership cannot, so it's normal for there to be raft leaders with no corresponding lease holder, but the reverse is not true and that's what we're trying to monitor. 
</details>
Nevermind, upon rereading the implementation it does look equivalent. I was confused by the new variable name which doesn't sound like it implies the `leaseCovers` requirement.

---

*[storage/store_metrics.go, line 34 \[r10\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/8861#-KQeEb55t1wu6MyIPbSV:-KQeEb55t1wu6MyIPbSW:b-xeh5js) ([raw file](https://github.com/cockroachdb/cockroach/blob/94a8c6a8797168c0500ae65014127da0799a45ee/storage/store_metrics.go#L34)):*
> ```Go
> 	metaRaftLeaderNotLeaseHolderCount = metric.Metadata{
> 		Name: "replicas.leaders_not_leaseholders",
> 		Help: "Total number of Replicas on this store that are Raft leaders but not Range lease holders.",
> ```

s/not Range lease holders/the range lease is held by another store/

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/8861)*
<!-- Sent from Reviewable.io -->
